core/src/type_info.rs

```diff
@@ -0,0 +1,304 @@
+            Schema::Atomic(Atomic::Scalar(BsonTypeName::Int)),
+            Nullability::NULLABLE
+        ),
+        MongoColMetadata::new(
+            "",
+            "".to_string(),
+            "FIXED_PREC_SCALE".to_string(),
+            Schema::Atomic(Atomic::Scalar(BsonTypeName::Int)),
+            Nullability::NO_NULLS
+        ),
+        MongoColMetadata::new(
+            "",
+            "".to_string(),
+            "AUTO_UNIQUE_VALUE".to_string(),
+            Schema::Atomic(Atomic::Scalar(BsonTypeName::Int)),
+            Nullability::NULLABLE
+        ),
+        MongoColMetadata::new(
+            "",
+            "".to_string(),
+            "LOCAL_TYPE_NAME".to_string(),
+            Schema::Atomic(Atomic::Scalar(BsonTypeName::String)),
+            Nullability::NULLABLE
+        ),
+        MongoColMetadata::new(
+            "",
+            "".to_string(),
+            "MINIMUM_SCALE".to_string(),
+            Schema::Atomic(Atomic::Scalar(BsonTypeName::Int)),
+            Nullability::NULLABLE
+        ),
+        MongoColMetadata::new(
+            "",
+            "".to_string(),
+            "MAXIMUM_SCALE".to_string(),
+            Schema::Atomic(Atomic::Scalar(BsonTypeName::Int)),
+            Nullability::NULLABLE
+        ),
+        MongoColMetadata::new(
+            "",
+            "".to_string(),
+            "SQL_DATA_TYPE".to_string(),
+            Schema::Atomic(Atomic::Scalar(BsonTypeName::Int)),
+            Nullability::NO_NULLS
+        ),
+        MongoColMetadata::new(
+            "",
+            "".to_string(),
+            "SQL_DATETIME_SUB".to_string(),
+            Schema::Atomic(Atomic::Scalar(BsonTypeName::Int)),
+            Nullability::NULLABLE
+        ),
+        MongoColMetadata::new(
+            "",
+            "".to_string(),
+            "NUM_PREC_RADIX".to_string(),
+            Schema::Atomic(Atomic::Scalar(BsonTypeName::Int)),
+            Nullability::NULLABLE
+        ),
+        MongoColMetadata::new(
+            "",
+            "".to_string(),
+            "INTERVAL_PRECISION".to_string(),
+            Schema::Atomic(Atomic::Scalar(BsonTypeName::Int)),
+            Nullability::NULLABLE
+        ),
+    ];
+}
+
+#[derive(Debug)]
+pub struct MongoTypesInfo {
+    current_type_index: isize,
+    sql_data_type: SqlDataType,
+}
+
+impl MongoTypesInfo {
+    pub fn new(sql_data_type: SqlDataType) -> MongoTypesInfo {
+        MongoTypesInfo {
+            current_type_index: 0,
+            sql_data_type,
+        }
+    }
+}
+
+impl MongoStatement for MongoTypesInfo {
+    // iterate through the list, searching for the next "valid" data type.
+    // a type is valid if its sql type matches the desired sql type, or if we are getting all types.
+    fn next(&mut self, _: Option<&MongoConnection>) -> Result<bool> {
+        loop {
+            self.current_type_index += 1;
+            if self.current_type_index > (DATA_TYPES.len() as isize)
```

Comment:
Ah this can be removed. I initially had `current_type_index` as an isize, initialized to -1, so that it could be used to directly index into `DATA_TYPES`, however turning it into a usize made the code more readable. This was a relic of that, that I can remove!